src/onnx/parse_attention.cpp

```diff
@@ -0,0 +1,967 @@
+        { // TODO: Add rotary embedding support -- parsed but not used right now
+            auto rotary_embedding_dim =
+                parser.parse_value(info.attributes.at("rotary_embedding_dim")).at<size_t>();
+
+            if(rotary_embedding_dim != 32 and rotary_embedding_dim != 64 and
+               rotary_embedding_dim != 128)
+            {
+                MIGRAPHX_THROW(
+                    "PARSE_ATTENTION: rotary_embedding_dim must be either 32, 64, or 128");
+            }
+
+            if(not attr_out.do_rotary)
+            {
+                MIGRAPHX_THROW(
+                    "PARSE_ATTENTION: rotary_embedding_dim must be used with do_rotary attribute");
+            }
+            attr_out.rotary_embedding_dim = rotary_embedding_dim;
+        }
+
+        if(contains(info.attributes, "scale"))
+        {
+            attr_out.scale = parser.parse_value(info.attributes.at("scale")).at<float>();
+            inferred_out.scale_not_query_sz = true;
+        }
+
+        if(contains(info.attributes, "unidirectional"))
+        {
+            attr_out.unidirectional =
+                (1 == parser.parse_value(info.attributes.at("unidirectional")).at<size_t>());
+            MIGRAPHX_THROW("PARSE_ATTENTION: unidirectional attr not supported");
+        }
+    }
+
+    // qkv values must be greater than zero to be "set"
+    static bool qkv_size_not_set(std::vector<size_t>& qkv_vec)
+    {
+        return std::any_of(qkv_vec.begin(), qkv_vec.end(), [](auto i) { return i <= 0; });
+    }
+
+    static void qkv_sizes_sum_arg_valid(const std::vector<size_t>& qkv_vec,
+                                        const instruction_ref input_arg,
+                                        const size_t dim,
+                                        const std::string& name)
+    {
+        if(std::accumulate(qkv_vec.begin(), qkv_vec.end(), size_t(0)) !=
+           input_arg->get_shape().lens().at(dim))
+        {
+            MIGRAPHX_THROW("Attention: q k v hidden sizes sum must match " + name + " tensor " +
+                           std::to_string(dim) + " dimension");
+        }
+    }
+
+    static bool weights_not_equal(const shape& weight_shape)
+    {
+        return (weight_shape.lens().at(1) % 3 != 0);
+    }
+
+    // simple call to check if the arg index exists
+    static bool check_and_return_arg(const std::vector<instruction_ref>& args,
+                                     const size_t index,
+                                     instruction_ref& output_arg)
+    {
+        if(args.size() > index)
+        {
+            output_arg = args.at(index);
+            return true;
+        }
+        return false;
+    }
+
+    static void handle_input(const instruction_ref& input_arg,
+                             const struct attention_attr& parsed_in,
+                             struct attention_inferred& inferred_out,
+                             std::vector<instruction_ref>& output_arg_vec)
```

Comment:
Return `instruction_ref` instead of using an output parameter. The vector can be appended in the caller.

Reply:
done